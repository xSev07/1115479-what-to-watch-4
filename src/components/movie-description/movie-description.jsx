import MovieNav from "../movie-nav/movie-nav.jsx";
import React from "react";
import MovieOverview from "../movie-overview/movie-overview.jsx";
import MovieDetails from "../movie-details/movie-details.jsx";
import MovieReviews from "../movie-reviews/movie-reviews.jsx";
import PropTypes from "prop-types";
import withActiveElement from "../../hocs/with-active-element/with-active-element.jsx";
import {MovieTab} from "../../const";
import {LoadError, Spinner} from "../svg/svg.jsx";

const getMovieReviewsTemplate = (loadingError, comments) => {
  if (loadingError) {
    return (
      <div style={{marginTop: `50px`, display: `flex`, justifyContent: `center`, alignItems: `center`}}>
        <h2 style={{color: `#d9cd8d`}}>Извините, у нашего сервера лапки. Попробуйте позднее</h2>
        <LoadError/>
      </div>
    );
  }
  if (!comments) {
    return (
      <div style={{marginTop: `50px`, display: `flex`, justifyContent: `center`, alignItems: `center`}}>
        <Spinner/>
      </div>
    );
  }
  return <MovieReviews comments={comments}/>;
};

const MovieDescription = (props) => {
  const {movie, comments, elements, activeElement, loadingCommentsError, onElementClick} = props;

  return (
    <div className="movie-card__desc">
      <MovieNav
        tabs={elements}
        activeTab={activeElement}
        onClick={onElementClick}
      />
      {activeElement === MovieTab.OVERVIEW && (
        <MovieOverview {...movie}/>
      )}
      {activeElement === MovieTab.DETAILS && (
        <MovieDetails {...movie}/>
      )}
      {activeElement === MovieTab.REVIEWS && (
        getMovieReviewsTemplate(loadingCommentsError, comments)
      )}
    </div>
  );
};

MovieDescription.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    votes: PropTypes.number.isRequired,
    producer: PropTypes.string.isRequired,
    actors: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    description: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    poster: PropTypes.string.isRequired,
    background: PropTypes.string.isRequired,
    backgroundColor: PropTypes.string.isRequired,
  }),
  comments: PropTypes.arrayOf(PropTypes.shape({
    commentId: PropTypes.string.isRequired,
    userId: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    date: PropTypes.instanceOf(Date).isRequired,
  })),
  elements: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  activeElement: PropTypes.string.isRequired,
  loadingCommentsError: PropTypes.bool.isRequired,
  onElementClick: PropTypes.func.isRequired,
};

export {MovieDescription};
export default withActiveElement(MovieDescription);
