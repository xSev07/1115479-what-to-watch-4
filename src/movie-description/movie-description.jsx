import MovieNav from "../components/movie-nav/movie-nav.jsx";
import React from "react";
import {getRatingTextDescription} from "../utils/common/common";

const MovieTab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

class MovieDescription extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      activeTab: MovieTab.OVERVIEW,
    };

    this._tabClickHandler = this._tabClickHandler.bind(this);
  }

  render() {
    const {votes, producer, rating, actors, description} = this.props;
    const textRating = getRatingTextDescription(rating);
    const actorsText = `${actors.join(`, `)} and other`;
    const DescriptionText = description.map((it, index) => <p key={index}>{it}</p>);
    const tabs = Object.values(MovieTab);

    return (
      <div className="movie-card__desc">
        <MovieNav
          tabs={tabs}
          activeTab={this.state.activeTab}
          onClick={this._tabClickHandler}
        />

        <div className="movie-rating">
          <div className="movie-rating__score">{rating}</div>
          <p className="movie-rating__meta">
            <span className="movie-rating__level">{textRating}</span>
            <span className="movie-rating__count">{votes} ratings</span>
          </p>
        </div>

        <div className="movie-card__text">
          {DescriptionText}

          <p className="movie-card__director"><strong>Director: {producer}</strong></p>

          <p className="movie-card__starring"><strong>Starring: {actorsText}</strong></p>
        </div>
      </div>
    );
  }

  _tabClickHandler(tab) {
    this.setState({activeTab: tab});
  }
}

export default MovieDescription;
