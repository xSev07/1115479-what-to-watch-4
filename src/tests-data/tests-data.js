export const movies = [
  {
    id: `Hotel-2000`,
    actors: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
    description: [
      `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
      `Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`
    ],
    duration: 60,
    genre: `Drama`,
    inList: false,
    producer: `Wes Andreson`,
    rating: 8.9,
    reviews: [],
    title: `The Grand Budapest Hotel`,
    videoPreview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    votes: 240,
    year: 2000,
  },
  {
    id: `Bohemian-2018`,
    actors: [
      `Rami Malek`,
      `Lucy Boynton`,
      `Gwilym Lee`
    ],
    description: [
      `The story of the legendary British rock band Queen and lead singer Freddie Mercury, leading up to their famous performance at Музыкальный фестиваль Live Aid (1985)`
    ],
    duration: 134,
    genre: `Drama`,
    inList: false,
    producer: `Bryan Singer`,
    rating: 8.0,
    reviews: [],
    title: `Bohemian Rhapsody`,
    videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    votes: 220,
    year: 2018,
  },
  {
    id: `Macbeth-2015`,
    actors: [
      `Michael Fassbender`,
      `Marion Cotillard`,
      `Paddy Considine`
    ],
    description: [
      `From the Academy Award winning producers of The King’s Speech and acclaimed director Justin Kurzel, comes a visceral and visually breath-taking retelling of the classic tale about an ambitious Scottish lord who seizes the throne with the help of his wife. Starring Academy Award nominee Michael Fassbender and Academy Award winner Marion Cotillard, Macbeth is both visually epic and a breathtaking experience.`
    ],
    duration: 113,
    genre: `Documental`,
    inList: false,
    producer: `Justin Kurzel`,
    rating: 6.6,
    reviews: [],
    title: `Macbeth`,
    videoPreview: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    votes: 220,
    year: 2015,
  },
  {
    id: `Aviator-2005`,
    actors: [
      `Leonardo DiCaprio`,
      `Cate Blanchett`,
      `Kate Beckinsale`
    ],
    description: [
      `The Aviator is a 2004 American epic biographical drama film directed by Martin Scorsese and written by John Logan. It stars Leonardo DiCaprio as Howard Hughes, Cate Blanchett as Katharine Hepburn, and Kate Beckinsale as Ava Gardner. The supporting cast features Ian Holm, John C. Reilly, Alec Baldwin, Jude Law as Errol Flynn, Gwen Stefani as Jean Harlow, Kelli Garner as Faith Domergue, Matt Ross, Willem Dafoe, Alan Alda, and Edward Herrmann.`,
      `Based on the 1993 non-fiction book Howard Hughes: The Secret Life by Charles Higham, the film depicts the life of Howard Hughes, an aviation pioneer and director of Hell's Angels. The film portrays his life from 1927–1947 during which time Hughes became a successful film producer and an aviation magnate while simultaneously growing more unstable due to severe obsessive–compulsive disorder (OCD).`
    ],
    duration: 170,
    genre: `Thriller`,
    inList: false,
    producer: `Martin Scorsese`,
    rating: 7.5,
    reviews: [],
    title: `Aviator`,
    videoPreview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    votes: 220,
    year: 2005,
  },
];

export const promoMovie = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  year: 2014
};
