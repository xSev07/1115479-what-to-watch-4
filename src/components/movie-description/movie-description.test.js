import React from "react";
import renderer from "react-test-renderer";
import {MovieDescription} from "./movie-description";
import {comments, movies} from "../../tests-data/tests-data";

describe(`Should MovieDescription render correctly`, () => {
  it(`Should render overview tab`, () => {
    const tree = renderer
      .create(
          <MovieDescription
            movie={movies[0]}
            comments={comments}
            elements={[`Overview`, `Details`, `Reviews`]}
            loadingCommentsError={false}
            activeElement={`Overview`}
            onElementClick={() => {
            }}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render details tab`, () => {
    const tree = renderer
      .create(
          <MovieDescription
            movie={movies[0]}
            comments={comments}
            elements={[`Overview`, `Details`, `Reviews`]}
            loadingCommentsError={false}
            activeElement={`Details`}
            onElementClick={() => {
            }}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render reviews tab`, () => {
    const tree = renderer
      .create(
          <MovieDescription
            movie={movies[0]}
            comments={comments}
            elements={[`Overview`, `Details`, `Reviews`]}
            loadingCommentsError={false}
            activeElement={`Reviews`}
            onElementClick={() => {
            }}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render reviews tab with loading error`, () => {
    const tree = renderer
      .create(
          <MovieDescription
            movie={movies[0]}
            comments={comments}
            elements={[`Overview`, `Details`, `Reviews`]}
            loadingCommentsError={true}
            activeElement={`Reviews`}
            onElementClick={() => {
            }}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should render reviews tab with waiting download comments`, () => {
    const tree = renderer
      .create(
          <MovieDescription
            movie={movies[0]}
            // comments={comments}
            elements={[`Overview`, `Details`, `Reviews`]}
            loadingCommentsError={false}
            activeElement={`Reviews`}
            onElementClick={() => {
            }}
          />
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
