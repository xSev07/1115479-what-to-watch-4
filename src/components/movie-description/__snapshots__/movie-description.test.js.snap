// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should MovieDescription render correctly Should render details tab 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-card__text movie-card__row"
  >
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Director
        </strong>
        <span
          className="movie-card__details-value"
        >
          Wes Andreson
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Starring
        </strong>
        <span
          className="movie-card__details-value"
        >
          Bill Murray
          , 
          <br />
          Edward Norton
          , 
          <br />
          Jude Law
          , 
          <br />
          Willem Dafoe
          , 
          <br />
        </span>
      </p>
    </div>
    <div
      className="movie-card__text-col"
    >
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Run Time
        </strong>
        <span
          className="movie-card__details-value"
        >
          1h 00m
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Genre
        </strong>
        <span
          className="movie-card__details-value"
        >
          drama
        </span>
      </p>
      <p
        className="movie-card__details-item"
      >
        <strong
          className="movie-card__details-name"
        >
          Released
        </strong>
        <span
          className="movie-card__details-value"
        >
          2000
        </span>
      </p>
    </div>
  </div>
</div>
`;

exports[`Should MovieDescription render correctly Should render overview tab 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-rating"
  >
    <div
      className="movie-rating__score"
    >
      8.9
    </div>
    <p
      className="movie-rating__meta"
    >
      <span
        className="movie-rating__level"
      >
        Very good
      </span>
      <span
        className="movie-rating__count"
      >
        240
         ratings
      </span>
    </p>
  </div>
  <div
    className="movie-card__text"
  >
    <p>
      In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
    </p>
    <p>
      Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.
    </p>
    <p
      className="movie-card__director"
    >
      <strong>
        Director: 
        Wes Andreson
      </strong>
    </p>
    <p
      className="movie-card__starring"
    >
      <strong>
        Starring: 
        Bill Murray, Edward Norton, Jude Law, Willem Dafoe and other
      </strong>
    </p>
  </div>
</div>
`;

exports[`Should MovieDescription render correctly Should render reviews tab 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-card__reviews movie-card__row"
  >
    <div
      className="movie-card__reviews-col"
    >
      <div
        className="review"
      >
        <blockquote
          className="review__quote"
        >
          <p
            className="review__text"
          >
            Test string
          </p>
          <footer
            className="review__details"
          >
            <cite
              className="review__author"
            >
              Test
            </cite>
            <time
              className="review__date"
              dateTime={2020-06-21T10:54:44.818Z}
            >
              June 21, 2020
            </time>
          </footer>
        </blockquote>
        <div
          className="review__rating"
        >
          5.2
        </div>
      </div>
    </div>
    <div
      className="movie-card__reviews-col"
    >
      <div
        className="review"
      >
        <blockquote
          className="review__quote"
        >
          <p
            className="review__text"
          >
            Test string 2
          </p>
          <footer
            className="review__details"
          >
            <cite
              className="review__author"
            >
              Test2
            </cite>
            <time
              className="review__date"
              dateTime={2020-05-11T10:54:44.818Z}
            >
              May 11, 2020
            </time>
          </footer>
        </blockquote>
        <div
          className="review__rating"
        >
          8
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Should MovieDescription render correctly Should render reviews tab with loading error 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "50px",
      }
    }
  >
    <h2
      style={
        Object {
          "color": "#d9cd8d",
        }
      }
    >
      Извините, у нашего сервера лапки. Попробуйте позднее
    </h2>
    <svg
      enableBackground="new 0 0 512 512"
      height="125"
      viewBox="0 0 512 512"
      width="100"
      x="0"
      xmlns="http://www.w3.org/2000/svg"
      y="0"
    >
      <path
        d="m375.1,432.2c-14.7,0-28.8-4.5-41.8-13.2-3.4-2.3-14.4-5.7-22.9,0-29.3,19.9-66.1,16.9-93-7.4-7.8-7-19.5-7-27.3,0-15,13.5-32.7,20.7-51.2,20.7-47.4,0-86-46.5-86-103.7 0-41.6 20.6-79 52.5-95.4 8.2-4.2 12.5-13.3 10.7-22.3-1.8-8.7-2.7-17-2.7-25.2 0-57.2 38.6-103.7 86-103.7 32.1,0 60.8,21.7 76.3,56.1 6,13.5 23.9,12.5 29,9.6 11.8-6.8 24.3-10.3 37.1-10.3 46.3,0 84.9,45.5 85.9,101.4 0.1,4.9 2,9.7 5.3,13.3 17.9,19.7 28.2,47.6 28.2,76.6-5.68434e-14,56.9-38.6,103.5-86.1,103.5zm93.1-202c-4.8-75.7-59.1-133.9-126.4-133.9-13.1,0-25.9,2.3-38.2,6.7-23.6-38.7-62-62-104.3-62-70,0-126.9,64.9-126.9,144.6 0,6.2 0.4,12.5 1.1,18.9-37.7,25.8-61.5,72.9-61.5,123.9 0,79.8 56.9,144.6 126.9,144.6 23.2,0 45.3-7 64.9-20.4 19.6,13.4 41.7,20.4 64.9,20.4 18.4,0 36.6-4.6 53.2-13.4 16.7,8.8 34.8,13.4 53.2,13.4 70,0 126.9-64.9 126.9-144.6 0-36.8-12-71.4-33.8-98.2z"
        fill="#d9cd8d"
      />
      <path
        d="m285.9,288.9l56.8-56.8c8-8 8-20.9 0-28.9-8-8-20.9-8-28.9,0l-56.8,56.8-56.8-56.8c-8-8-20.9-8-28.9,0-8,8-8,20.9 0,28.9l56.8,56.8-56.8,56.8c-8,8-8,20.9 0,28.9 8,8 20.9,8 28.9,0l56.8-56.8 56.8,56.8c8,8 20.9,8 28.9,0 8-8 8-20.9 0-28.9l-56.8-56.8z"
        fill="#d9cd8d"
      />
    </svg>
  </div>
</div>
`;

exports[`Should MovieDescription render correctly Should render reviews tab with waiting download comments 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginTop": "50px",
      }
    }
  >
    <svg
      display="block"
      height="200"
      preserveAspectRatio="xMidYMid"
      style={
        Object {
          "margin": "auto",
        }
      }
      viewBox="0 0 100 100"
      width="200"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M85.141 70.133a40 40 0 00-71.194-36.487A42 40-62.865 0185.14 70.133"
        fill="#d9cd8d"
      >
        <animateTransform
          attributeName="transform"
          dur="1s"
          keyTimes="0;1"
          repeatCount="indefinite"
          type="rotate"
          values="0 50 51;360 50 51"
        />
      </path>
    </svg>
  </div>
</div>
`;
